Added JSON-based changes to auto-generate the changelog.

Alexhuszagh opened the pull request:
Solves merge conflicts between changelogs, by using sorting and type-based merging for the changelog from a series of JSON files.

Closes #662.

Emilgardis commented on .changes/README.md:
````diff
@@ -0,0 +1,23 @@
+# Changes
+
+This stores changes to automatically generate the changelog, to avoid merge conflicts. Files should be in a JSON format, with the following format:
+
+```json
+{
+    "description": "single-line description to add to the CHANGELOG.",
+    "issues": [894],
+    "type": "added",
+    "breaking": false
+}
+```
````
can we add a `template.json` file with this, probably without the optional breaking

Alexhuszagh replied:
Sure, guess we just ignore it?

Alexhuszagh replied:
Done, I've added 2 files, one for an array and one for an object-based change. I'll add them to the README as well.

Alexhuszagh commented:
This [action](https://github.com/marketplace/actions/get-all-changed-files) should be able to resolve our changelog check, btw.

We also, if the draft is not used, need to delete every JSON file in `.changes`.

Emilgardis commented:
could this handle the situations when there are multiple prs before release, i.e entries adjusting the same changelog entry?

Don't want the changelog to be a per-pr changelog, for that we could just use autogenerated changelogs

Alexhuszagh commented:
> could this handle the situations when there are multiple prs before release doing the same thing, i.e entries adjusting the same changelog entry?
> 
> Don't want the changelog to be a per-pr changelog, for that we could just use autogenerated changelogs

Absolutely. The one thing it doesn't do right now is handle multiple entries having the same PR number. That also needs to be, preferably without changing the syntax. An example where this would fail is #905.

Alexhuszagh commented:
> could this handle the situations when there are multiple prs before release doing the same thing, i.e entries adjusting the same changelog entry?
> 
> Don't want the changelog to be a per-pr changelog, for that we could just use autogenerated changelogs

The way to modify the entries pre-release would be to just modify that JSON file I believe. I've also added it so a single PR can modify multiple different changes: you can pass an object or array.

```json
[
    {
        "description": "sample 1, sample 2",
        "type": "added"
    },
    {
        "description": "sample 3, sample 4",
        "type": "added"
    },
    {
        "description": "some internal stuff",
        "issues": [662],
        "type": "internal"
    }
]
```

Or

```json
{
    "description": "use JSON-based files to autogenerate CHANGELOG.md",
    "issues": [662],
    "type": "internal"
}
```

This will get merged to:

```markdown
### Added

- #919 - sample 1, sample 2
- #919 - sample 3, sample 4

### Internal

- #919 - some internal stuff
- #918 - use JSON-based files to autogenerate CHANGELOG.md
```

To modify an entry prior to release, just modify the JSON entry. It handles any number of changes, and doesn't use the changes until release (when the pre-release hook is used). However, you can update the CHANGELOG at any time, and the draft is meant to do that.

Alexhuszagh commented:
Ok I've updated the Github Actions for the changelog check to:
- Check if any file has changed in the `.changes/*.json` directory (added, modified, or removed).
- Validate every JSON file by parsing using our serde deserializer, if added or modified.
- Validate the CHANGELOG.md file (but don't look for changes).
- Added a pre-release hook that auto-generates the update CHANGELOG.md

This should do everything. How it works is quite simple:
1. Parses every entry in `.changes/*.json` if they are not template files.
2. The filename is the PR number, the contents are the CHANGELOG data (description and type are required).
3. These are then added to vecs by each type (Added, Changed, Fixed, Internal, Removed).
4. The CHANGELOG.md file is read, and the contents between `<!-- next-header -->` and `<!-- last-header -->` are read.
5. Each section that starts with a `###` is used to identify the changelog type, and the PR number and description our extracted from each entry starting with `- `.
6. These are then added to different vecs by each type (Added, Changed, Fixed, Internal, Removed).
7. These two sets of vecs are merged and sorted by PR number.
8. These are then formatted to string, and the header (before `<!-- next-header -->`) and the footer (after `<!-- last-header -->`) are printed verbatim.

This means there will never be conflicts: multiple PRs can have the same types (numerous added types, etc.), can have one or more types (see the example), and since they are merged into a vector and sorted by PR number, the result will always be in descending order and look nice.

Emilgardis commented on CHANGELOG.md:
```diff
@@ -33,6 +33,8 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 - #868 - ignore the `CARGO` environment variable.
 - #867 - fixed parsing of `build.env.passthrough` config values.
 
+<!-- last-header -->
+
```
the changelog needs to have a mention of `.changesets` since its no longer a live document, `.changes` is

Alexhuszagh replied:
This has now been documented as follows:

```markdown
# Change Log

All notable changes to this project will be documented in this file. This is an automatically-generated document: entries are added via changesets present in the `.changes` directory.
This project adheres to [Semantic Versioning](http://semver.org/).
```

Emilgardis commented on .changes/README.md:
```diff
@@ -0,0 +1,46 @@
+- removed (Removed)
+- internal (Internal)
+
+`breaking` is optional and defaults to false. if `breaking` is present for any active changes, we will not allow a minor (or pre-1.0) patch version to be generated. `issues` is also optional, and is currently unused, and is an array of issues fixed by the PR, and defaults to an empty array.
```
This `breaking` comment doesn't seem to be implemented.

Alexhuszagh replied:
It isn't, neither is `issues` right now. I'll remove it from the documentation probably. The idea was nice, but it's not reported in the generated CHANGELOG and is probably too noisy to be useful.

Emilgardis commented on .changes/README.md:
```diff
@@ -0,0 +1,46 @@
+- removed (Removed)
+- internal (Internal)
+
+`breaking` is optional and defaults to false. if `breaking` is present for any active changes, we will not allow a minor (or pre-1.0) patch version to be generated. `issues` is also optional, and is currently unused, and is an array of issues fixed by the PR, and defaults to an empty array.
```
I think if breaking is present, a `BREAKING:` header should probably be added to the body of the description

Alexhuszagh replied:
Or actually, that's a much better solution. Nice.

Emilgardis commented on xtask/src/changelog.rs:
```diff
@@ -0,0 +1,369 @@
+use serde::Deserialize;
+
+#[derive(Args, Debug)]
+pub struct BuildChangelog {
+    /// Provide verbose diagnostic output.
+    #[clap(short, long)]
+    pub verbose: bool,
+    /// Do not print cross log messages.
+    #[clap(short, long)]
+    pub quiet: bool,
+    /// Whether messages should use color output.
+    #[clap(long)]
+    pub color: Option<String>,
+    /// Build a draft changelog (to `CHANGELOG.md.draft`).
+    #[clap(short, long)]
+    draft: bool,
+}
```
this commands need to support `cargo release <v>`, which is a dry-run.

It should also take into account if the release is a pre-release, and can also take the new version and previous version to see if it's a breaking change, and error if there is a breaking change but not a breaking bump.


`cargo-release` defines a [few environment variables](https://github.com/crate-ci/cargo-release/blob/master/docs/reference.md#hook-environment-variables) that can help

Emilgardis reviewed (changes requested):
looks good!

The `./.github/actions/cargo-publish` action needs to be updated to generate the changelog.

There's also an issue with pre-releases, should pre-releases delete the changeset files or not?

Alexhuszagh commented on .changes/README.md:
```diff
@@ -0,0 +1,46 @@
+
+`breaking` is optional and defaults to false. if `breaking` is present for any active changes, we will not allow a minor (or pre-1.0) patch version to be generated. `issues` is also optional, and is currently unused, and is an array of issues fixed by the PR, and defaults to an empty array.
+
+The file numbers should be `${pr}.json`
+
+If multiple changes are made in a single PR, you can also pass an array of entries:
```
> I'd like it to be possible to have multiple prs on one line, and also an entry without pr number. e.g like the two below entries in 0.2.2
> 
> ```
> - #767, #788 - added the `cross-util` and `xtask` commands.
> - Change rust edition to 2021 and bump MSRV for the cross binary to 1.58.1
> ```
> 
> this could be implemented by naming the file `767-788.json` or similar, and for regular entries without a pr attached `_901.json` maybe?

That should definitely be doable, we just need a consistent naming convention. I guess we'd have to make a sort key and then make the PR optional, which could be an option or an issue number, and one must be present.

Alexhuszagh replied:
This has been fixed with the following: entries with the format `issue${issue}.json` won't have an associated PR number, and the rest will. For example, a breaking change without a PR attached would produce something like `- BREAKING: this is one added entry.`.

It supports the following syntax for PRs:

- `${pr1}-${pr2}(...).json`, which will be output as `- #979,981 - this has 2 PRs associated.`
- `${issue1}-${issue2}.json`, which will be output as `- this is one added entry.`

Both PRs and issues can have 1 or more associated numbers with it, and both can have an object or an array objects for each entry. See the unittests for extensive documentation on this.

Alexhuszagh commented on .changes/README.md:
```diff
@@ -0,0 +1,46 @@
+
+`breaking` is optional and defaults to false. if `breaking` is present for any active changes, we will not allow a minor (or pre-1.0) patch version to be generated. `issues` is also optional, and is currently unused, and is an array of issues fixed by the PR, and defaults to an empty array.
+
+The file numbers should be `${pr}.json`
```
It's not too bad I think, if it has the issue number since we can just choose to report those. We would need to write them at the end, however, since they won't sort well.

Alexhuszagh replied:
Done: PR numbers are now optional, and they are added to the bottom, sorted by description.

Alexhuszagh commented:
I think it's probably a good idea, in retrospect, not delete the `.changes`, but to move them to a dedicated subdirectory based on the release number? Then the default is to make a draft, and if the release is present, move all files over and then update the CHANGELOG.

Alexhuszagh commented:
I believe the only thing left is adding the `cargo-publish`.

You can test the parsing and generating of the CHANGELOG via `cargo test --workspace`, which auto-runs the tests to parse and format the documents, as well as parse and validate the deserializers from JSON. It tests having no PR, multiple PRs, multiple issues, breaking changes, and no breaking changes. For example, the files it tests that parse correctly are:

https://github.com/Alexhuszagh/cross/blob/c20f3f0855c003559ec448a3ea4b6081ee6bdbe9/xtask/src/changelog.rs#L821-L852

And it checks that the output is reasonable, say, for here:

https://github.com/Alexhuszagh/cross/blob/c20f3f0855c003559ec448a3ea4b6081ee6bdbe9/xtask/src/changelog.rs#L799-L816

However, `cargo release` isn't working correctly yet, since the hook is run **after** the pre-release statements are made, which I need to fix.

Emilgardis commented:
This is great, however I feel like it should actually probably be part of its own binary.

Maybe something for https://github.com/crate-ci/ ?

Alexhuszagh commented:
This now handles `cargo-release` properly (well, it should, assuming the pre-release replacements are actually written to file prior to the hook getting invoked, which I believe they are). The only issue right now is it also makes the changes if the release is a release-candidate. I believe we shouldn't update the changelog version in that case, just add the changes but keep them in an `Unreleased` state.

You can test it for yourself running `cargo release patch --verbose`.

Alexhuszagh commented:
This finally handles pre-releases correctly, and it will behave as follows:

When using a pre-release, it will update `CHANGELOG.md` but keep everything in an `Unreleased` section. The changes will be moved to `.changes/{major}.{minor}.{patch}`, so they will be in the same directory as any additional changes prior to the release.

If a release occurs, it will change the unreleased section from:

```markdown
## [Unreleased] - ReleaseDate

### Internal

- #918 - use JSON-based files to autogenerate CHANGELOG.md
```

to:

```markdown
## [Unreleased] - ReleaseDate

## [v0.2.4] - 2022-07-10

### Internal

- #918 - use JSON-based files to autogenerate CHANGELOG.md
```

Emilgardis commented:
looks good! There's something wrong with the behaviour though.

```
cross on changesets [$] is 📦 v0.2.3 via 🦀 v1.61.0 
❯ cargo release 0.3.0 --execute
[2022-07-10T12:54:01Z WARN  cargo_release::git] Push target `origin/changesets` doesn't exist
Release cross 0.3.0? [y/N] 
y
[2022-07-10T12:54:03Z INFO  cargo_release::release] Update cross to version 0.3.0
   Compiling cross v0.3.0 (/Users/emil/workspace/cross)
   Compiling xtask v0.0.0-dev.0 (/Users/emil/workspace/cross/xtask)
    Finished dev [unoptimized + debuginfo] target(s) in 2.90s
     Running `target/debug/xtask build-changelog`
Building the changelog.
Running rustfmt and clipp sy checks.
+ flags=(--all-features --all-targets --workspace)
+ cargo fmt -- --check
+ cargo +nightly
+ cargo +nightly clippy --all-features --all-targets --workspace -- --deny warnings
    Blocking waiting for file lock on build directory
   Compiling cross v0.3.0 (/Users/emil/workspace/cross)
    Checking xtask v0.0.0-dev.0 (/Users/emil/workspace/cross/xtask)
    Finished dev [unoptimized + debuginfo] target(s) in 4.60s
[changesets cf3b8ee] release version 0.3.0
 4 files changed, 11 insertions(+), 8 deletions(-)
 delete mode 100644 .changes/918.json

cross on changesets [$?] is 📦 v0.3.0 via 🦀 v1.61.0 took 11s 
❯ git status
On branch changesets
Untracked files:
  (use "git add <file>..." to include in what will be committed)
        .changes/0.3.0/

nothing added to commit but untracked files present (use "git add" to track)
```

the `0.3.0` folder does not get added. Also, should we really keep these files after release?

Alexhuszagh commented:
> the `0.3.0` folder does not get added. Also, should we really keep these files after release?

I've changed from renaming those files to just deleting them. This is ready for review.

Emilgardis commented:
I think json is fine for this

Emilgardis commented on xtask/src/changelog.rs:
```diff
@@ -0,0 +1,913 @@
+fn delete_changes(root: &Path, version: &str) -> cross::Result<()> {
+    // move all files to the denoted version release
+    let src = root.join(".changes");
+    let dst = src.join(Path::new(&as_release(version)?));
+    fs::create_dir_all(&dst)?;
```
this can be removed

Emilgardis reviewed (approved):
lgtm!

fix the trailing folder and we can start to use this!

Alexhuszagh commented:
bors r=Emilgardis
